clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/core/ProjectClustersService.java

```diff
@@ -0,0 +1,415 @@
+        continue;
+      }
+
+      projectClusters.put(project.name, getProjectClusters(project));
+    }
+
+    return projectClusters;
+  }
+
+  public List<ClusterModel> getProjectClusters(String projectName) {
+    Map projectData = front50Service.getProject(projectName);
+
+    if (projectData == null) {
+      return null;
+    }
+
+    Project project;
+    try {
+      project = objectMapper.convertValue(projectData, Project.class);
+    } catch (IllegalArgumentException e) {
+      throw new MalformedProjectDataException("Could not marshal project to internal model: " + projectName, e);
+    }
+
+    return getProjectClusters(project);
+  }
+
+  public List<ClusterModel> getProjectClusters(Project project) {
+    List<String> applicationsToRetrieve = Optional.ofNullable(project.config.applications)
+      .orElse(Collections.emptyList());
+    Map<String, Set<Cluster>> allClusters = retrieveClusters(applicationsToRetrieve);
+
+    return project.config.clusters.stream()
+      .map(projectCluster -> {
+        List<String> applications = Optional.ofNullable(projectCluster.applications).orElse(project.config.applications);
+        List<ApplicationClusterModel> applicationModels = applications.stream()
+          .map(application -> {
+            Set<Cluster> appClusters = allClusters.get(application);
+            Set<Cluster> clusterMatches = findClustersForProject(appClusters, projectCluster);
+            return new ApplicationClusterModel(application, clusterMatches);
+          })
+          .collect(Collectors.toList());
+
+        return new ClusterModel(
+          projectCluster.account,
+          projectCluster.stack,
+          projectCluster.detail,
+          applicationModels
+        );
+      })
+      .collect(Collectors.toList());
+  }
+
+  private Map<String, Set<Cluster>> retrieveClusters(List<String> applications) {
+    Map<String, Set<Cluster>> allClusters = new HashMap<>();
+
+    for (String application : applications) {
+      for (RetrievedClusters clusters : retrieveApplication(application)) {
+        allClusters.computeIfAbsent(clusters.application, s -> new HashSet<>())
+          .addAll(clusters.clusters);
+      }
+    }
+
+    return allClusters;
+  }
+
+  private Set<Cluster> findClustersForProject(Set<Cluster> appClusters, ProjectCluster projectCluster) {
+    if (appClusters == null || appClusters.isEmpty()) {
+      return Collections.emptySet();
+    }
+
+    return appClusters.stream()
+      .filter(appCluster -> {
+        Names clusterNameParts = Names.parseName(appCluster.getName());
+        return appCluster.getAccountName().equals(projectCluster.account) &&
+          nameMatches(clusterNameParts.getStack(), projectCluster.stack) &&
+          nameMatches(clusterNameParts.getDetail(), projectCluster.detail);
+      })
+      .collect(Collectors.toSet());
+  }
+
+  private List<RetrievedClusters> retrieveApplication(String application) {
```

Comment:
might just be me but I find the `retrieveApplication` name a bit confusing. If it's retrieving cluster for an application, `retrieveClusters(String application)` makes more sense to me

Reply:
This is all just copied from the old stuff / will be rewritten when Adam gets to it, but will change.